tokenizers/src/decoders/wordpiece.rs

```diff
@@ -44,12 +44,21 @@ pub fn cleanup(dirty_input: String) -> String {
 }
 
 impl Decoder for WordPiece {
-    fn decode(&self, tokens: Vec<String>) -> Result<String> {
-        let mut output = tokens.join(" ").replace(&format!(" {}", self.prefix), "");
-        if self.cleanup {
-            output = cleanup(output);
-        }
-
-        Ok(output)
+    fn decode(&self, mut tokens: Vec<String>) -> Result<Vec<String>> {
+        tokens
+            .iter_mut()
+            .enumerate()
+            .map(|(i, token)| {
+                if token.starts_with(&self.prefix) {
+                    *token = token.replacen(&self.prefix, "", 1);
+                } else if i != 0 {
+                    *token = format!(" {}", token);
+                }
+                if self.cleanup {
+                    *token = cleanup(token.clone());
```

Comment:
Isn't better API-wise to pass in a reference, `&token` and call a `.to_owned()` or `.clone()` inside the `cleanup()` method ?
Feels cleaner, the responsibility of the clone is imo `cleanup`'s.

Reply:
I think it's almost always preferable to not clone on a user's behalf and let the user decide how best to pass owned instances (if you really require ownership).

It seems maybe this function doesn't actually require ownership, I'll change to reflect that.

Reply:
I see what you mean. W/out the clone, you will have to create a new instance of a `String`, which is somehow similar to cloning ^^

Reply:
If that is the rust way, so be it 😄